src/ng/directive/as.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+/**
+ * @ngdoc directive
+ * @name as
+ * @restrict A
+ *
+ * @description
+ * Publishes a component controller or a DOM element into the current scope.
+ *
+ * @element ANY
+ * @param {expression} as {@link guide/expression Expression} to assign the controller.
+ *
+ * This directive allows to access to the controller of components in your
+ * current template or from current controller.
+ *
+ * @example
+ * ### Access to component controller
+ * This example demonstrates basic publishing of the component controller into the scope.
+ * <example name="asDirectiveComponentExample" module="asComponentExample">
+ *   <file name="index.html">
+ *     <toggle as="myToggle"></toggle>
+ *     <button ng-click="myToggle.toggle()">Toggle</button>
+ *     <div ng-show="myToggle.isOpen()">You are using a children component to show it.</div>
+ *   </file>
+ *   <file name="script.js">
+ *     angular.module('asComponentExample', [])
+ *      .component('toggle', {
+ *         controller: function() {
+ *           var opened = false;
+ *           this.isOpen = function() { return opened; };
+ *           this.toggle = function() { opened = !opened; };
+ *         }
+ *       });
+ *   </file>
+ *   <file name="protractor.js" type="protractor">
+ *      it('should publish the toggle into the scope', function() {
+ *        var toggle = element(by.buttonText('Toggle'));
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(false);
+ *        toggle.click();
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(true);
+ *      });
+ *   </file>
+ * </example>
+ */
+var componentAsDirective = ['$parse', function($parse) {
+  return {
+    restrict: 'A',
+    compile: function(tElement, tAttrs) {
+      // gets the expected controller name, converts <x-some-thing> into "someThing")
+      var controllerName = tAttrs.$normalize(tElement[0].tagName.toLowerCase())
+        .replace(/[^\w]\w/g, function(a) { return a.slice(1).toUpperCase(); });
+
+      // special case: ngForm is the only known directive that uses an undocumented
+      // feature of .directive able to change the name of the directive
+      if (controllerName === 'ngForm') {
+        controllerName = 'form';
+      }
+
+      // get the setter for the as attribute
+      var setter = $parse(tAttrs.as).assign;
+
+      return function(scope, iElement, iAttrs) {
+        // gets the controller of the current element (see jqLiteController for details)
+        var controller = iElement.data('$' + controllerName + 'Controller');
```

Comment:
Yea, don't use `jqLiteData` because as you found out, that uses the internal jqLite data store at all times.  If instead you use `jqLite.data`, it will use the appropriate storage if jQuery is loaded since jqLite is replaced with jQuery in those cases.